rustls/src/client/client_conn.rs

```diff
@@ -386,6 +389,93 @@ pub enum Tls12Resumption {
+///
+/// Note: differs from the protocol-encoded EchConfig (`EchConfigMsg`).
+#[derive(Clone, Debug)]
+#[allow(dead_code)] // TODO(@cpu): remove in subsequent commit.
+pub struct EchConfig {
+    /// The provider to use for HPKE operations.
+    pub(crate) hpke_provider: &'static dyn HpkeProvider,
+
+    /// The selected EchConfig.
+    pub(crate) config: EchConfigMsg,
+
+    /// An HPKE suite from the `config` we have selected as compatible with the `hpke_provider`.
+    pub(crate) suite: HpkeSuite,
+}
+
+impl EchConfig {
+    /// Construct an EchConfig by selecting a ECH config from the provided bytes that is compatible
+    /// with the given HPKE provider.
+    ///
+    /// The config bytes should be sourced from a DNS-over-HTTPS lookup resolving the `HTTPS` service
+    /// record record type for the host name of the server you wish to connect to using ECH,
+    /// and extracting the ECH service record parameter.
+    ///
+    /// One of the provided configurations must be compatible with the HPKE provider's supported
+    /// suites or an error will be returned.
+    // TODO(@cpu): update docstring with a pointer to the example program using hickory-dns.
+    // TODO(@cpu): consider dedicated HpkeError type(s).
+    pub fn new(
+        ech_configs: impl AsRef<[u8]>,
+        hpke_provider: &'static dyn HpkeProvider,
+    ) -> Result<Self, Error> {
+        // Try to unmarshal the ECH config bytes as a list of EchConfig structs, or a single EchConfig
+        // TODO(@cpu): verify if this is needed or if we should always assume a Vec...
+        let ech_configs = match Vec::<EchConfigMsg>::read(&mut Reader::init(ech_configs.as_ref())) {
+            Ok(configs) => configs,
+            Err(_) => {
+                vec![EchConfigMsg::read(&mut Reader::init(ech_configs.as_ref()))?]
+            }
+        };
+
+        let (config, suite) = ech_configs
+            .iter()
+            .find_map(|config| {
+                if config.version != EchVersion::V14 {
+                    return None; // Unsupported version.
+                }
+
+                let contents = &config.contents;
+                if contents.has_unknown_extension() || contents.has_duplicate_extension() {
+                    return None; // Unsupported, or malformed extensions.
+                }
+
+                let key_config = &contents.key_config;
+                for cipher_suite in &key_config.symmetric_cipher_suites {
+                    // TODO(@cpu): Confirm we known the tag length for the chosen AEAD ID.
+                    let suite = HpkeSuite {
+                        kem: key_config.kem_id,
+                        sym: *cipher_suite,
+                    };
+                    if hpke_provider.supports_suite(&suite) {
+                        return Some((config.clone(), suite));
+                    }
+                }
+
+                None
+            })
+            .ok_or_else(|| Error::General("No compatible ECH config".into()))?;
```

Comment:
Style nit: I think this might be clearer as a `for`-loop or a `let (config, suite) = loop { .. }`.

Reply:
I pulled this out into a separate fn that uses a for loop.